pkg/provisioner/templates/containerd.go

```diff
@@ -32,29 +32,121 @@ const containerdTemplate = `
+if [ -z "$CONTAINERD_VERSION" ]; then
+    echo "Fetching latest stable containerd version..."
+    CONTAINERD_VERSION=$(curl -fsSL https://api.github.com/repos/containerd/containerd/releases/latest | grep '"tag_name":' | cut -d '"' -f 4 | sed 's/v//')
 
-# Install containerd
-install_packages_with_retry containerd.io=${CONTAINERD_VERSION}-1
+    if [ -z "$CONTAINERD_VERSION" ]; then
+        echo "Failed to fetch latest Containerd version. Exiting."
+        exit 1
+    fi
+fi
+echo "Using containerd version: $CONTAINERD_VERSION"
 
-# Configure containerd and start service
-mkdir -p /etc/containerd
+ARCH=$(uname -m)
+CONTAINERD_TAR="containerd-${CONTAINERD_VERSION}-linux-${ARCH}.tar.gz"
+CONTAINERD_URL="https://github.com/containerd/containerd/releases/download/v${CONTAINERD_VERSION}/${CONTAINERD_TAR}"
+CONTAINERD_SHA256_URL="https://github.com/containerd/containerd/releases/download/v${CONTAINERD_VERSION}/SHA256SUMS"
+
+echo "Downloading and extracting containerd ${CONTAINERD_VERSION} from: $CONTAINERD_URL"
+
+# Download containerd tarball and checksum
+curl -fsSL -o ${CONTAINERD_TAR} ${CONTAINERD_URL}
+curl -fsSL -o containerd_SHA256SUMS ${CONTAINERD_SHA256_URL}
+
+# Verify SHA256 checksum
+EXPECTED_CHECKSUM=$(grep ${CONTAINERD_TAR} containerd_SHA256SUMS | awk '{print $1}')
+ACTUAL_CHECKSUM=$(sha256sum ${CONTAINERD_TAR} | awk '{print $1}')
+
+if [ "$EXPECTED_CHECKSUM" != "$ACTUAL_CHECKSUM" ]; then
+    echo "Checksum verification failed for containerd. Exiting."
+    exit 1
+fi
+
+# Stream directly into tar to avoid saving the archive
+cat ${CONTAINERD_TAR} | sudo tar Cxzvf /usr/local -
+
+# Cleanup
+rm -f ${CONTAINERD_TAR} containerd_SHA256SUMS
+
+echo "Containerd ${CONTAINERD_VERSION} installed successfully."
+
+# Fetch latest stable RUNC version from GitHub
+echo "Fetching latest stable runc version..."
+RUNC_VERSION=$(curl -fsSL https://api.github.com/repos/opencontainers/runc/releases/latest | grep '"tag_name":' | cut -d '"' -f 4 | sed 's/v//')
+
+if [ -z "$RUNC_VERSION" ]; then
+    echo "Failed to fetch latest RUNC version. Using default version."
+    RUNC_VERSION="1.2.6"
+fi
+
+RUNC_URL="https://github.com/opencontainers/runc/releases/download/v${RUNC_VERSION}/runc.${ARCH}"
+RUNC_SHA256_URL="https://github.com/opencontainers/runc/releases/download/v${RUNC_VERSION}/SHA256SUMS"
+
+echo "Downloading runc ${RUNC_VERSION} from: $RUNC_URL"
+
+# Download runc binary and checksum
+curl -fsSL -o runc.${ARCH} ${RUNC_URL}
+curl -fsSL -o SHA256SUMS ${RUNC_SHA256_URL}
+
+# Verify SHA256 checksum
+EXPECTED_CHECKSUM=$(grep "runc.${ARCH}" SHA256SUMS | awk '{print $1}')
+ACTUAL_CHECKSUM=$(sha256sum runc.${ARCH} | awk '{print $1}')
+
+if [ "$EXPECTED_CHECKSUM" != "$ACTUAL_CHECKSUM" ]; then
+    echo "Checksum verification failed for runc. Exiting."
+    exit 1
+fi
+
+sudo install -m 755 runc.${ARCH} /usr/local/sbin/runc
+rm -f runc.${ARCH} SHA256SUMS
+
+echo "Runc ${RUNC_VERSION} installed successfully."
+
+# Install CNI plugins
+CNI_VERSION="1.1.1"
```

Comment:
Any particular reason that we don't handle this version in the same way as the other packages?